SG14/inplace_function.h

```diff
@@ -29,340 +29,331 @@
-	template<typename CallableT>
-	inplace_function& operator=(const CallableT& target)
-	{
-		this->clear();
-		this->set(target);
-		return *this;
-	}
-
-	// Assign a new target by way of moving
-	// If the move constructor of target object throws, this is left in uninitialized state
-	template<typename CallableT, class = typename std::enable_if<!std::is_lvalue_reference<CallableT>::value>::type>
-	inplace_function& operator=(CallableT&& target)
-	{
-		this->clear();
-		this->set(std::move(target));
-		return *this;
-	}
-
-	// Compares this inplace function with a null pointer
-	// Empty functions compare equal, non-empty functions compare unequal
-	constexpr bool operator ==(std::nullptr_t) noexcept
-	{
-		return !operator bool();
-	}
-
-	// Compares this inplace function with a null pointer
-	// Empty functions compare equal, non-empty functions compare unequal
-	constexpr bool operator !=(std::nullptr_t) noexcept
-	{
-		return operator bool();
-	}
-
-	// Converts to 'true' if assigned
-	explicit constexpr operator bool() const noexcept
-	{
-		return m_InvokeFctPtr != &DefaultFunction;
-	}
-
-	// Invokes the target
-	// Throws std::bad_function_call if not assigned
-	RetT operator () (ArgsT... args) const
-	{
-		return m_InvokeFctPtr(std::forward<ArgsT>(args)..., data());
-	}
-
-	// Swaps two targets
-	void swap(inplace_function& other)
-	{
-		BufferType tempData;
-		this->move(m_Data, tempData);
-		other.move(other.m_Data, m_Data);
-		this->move(tempData, other.m_Data);
-		std::swap(m_InvokeFctPtr, other.m_InvokeFctPtr);
-		std::swap(m_ManagerFctPtr, other.m_ManagerFctPtr);
-	}
+  using capacity = std::integral_constant<size_t, Capacity>;
+  using alignment = std::integral_constant<size_t, Alignment>;
+
+  using storage_t = std::aligned_storage_t<Capacity, Alignment>;
+  using vtable_t = detail::vtable<R, Args...>;
+  using vtable_ptr_t = const vtable_t*;
+
+  template <typename, size_t, size_t>	friend class inplace_function;
+
+  inplace_function() noexcept :
+    vtable_ptr_{std::addressof(detail::empty_vtable<R, Args...>)}
+  {}
+
+  template<
+    typename T,
+    typename C = std::decay_t<T>,
+    typename = std::enable_if_t<
+      !(std::is_same<C, inplace_function>::value
+      || std::is_convertible<C, inplace_function>::value)
+    >
+  >
+  inplace_function(T&& closure)
+  {
+    static_assert(std::is_invocable_r<R, C, Args...>::value,
+      "Function closure has to be invocable"
+    );
+
+    static_assert(std::is_copy_constructible<C>::value,
+      "Constructing function with move only type is invalid"
+    );
+
+    static_assert(sizeof(C) <= Capacity, "Inplace function closure too large");
+
+    static_assert(
+      Alignment % std::alignment_of<C>::value == 0,
+      "Incompatible function closure alignment"
+    );
+
+    static const vtable_t vt{detail::wrapper<C>{}};
+    vtable_ptr_ = std::addressof(vt);
+
+    new (std::addressof(storage_)) C{std::forward<C>(closure)};
```

Comment:
Here `std::forward<C>` should be `std::forward<T>`. I wonder if it's possible to write a test case for this?

Reply:
Good spotting. The typename C is a bit confusing, given that it could be capacity or closure.